### PR TITLE
TDD project

### PR DESCRIPTION
**Pull request by Dangelo-JAN**
# We created a TDD projejct: TDD project

In this milestoone we:

- Create a class called `Solver`.
- Create a method called `factorial` that takes one argument, an integer `N`, and returns the [factorial](https://en.wikipedia.org/wiki/Factorial) for that number.
- Create a method called `reverse` that takes one argument, a string `word`, and returns `word` reversed (e.g. if `word` is `"hello"` it returns `"olleh"`).
- Create a method called `fizzbuzz` that takes one argument, an integer `N`, and returns a string. The returned string is constructed following these rules:
  - When `N` is divisible by 3, return `"fizz"`.
  - When `N` is divisible by 5, return `"buzz"`.
  - When `N` is divisible by 3 and 5, return `"fizzbuzz"`.

### CODE REVIEW

File: tdd_project.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Solver
+  def factorial(num)
+    return nil unless num.is_a? Integer
+    return 1 if num <= 0
+
+    num * factorial(num - 1)
+  end
```
Comment (Lameck1) [inline tdd_project.rb]: - [ ] You've done well by writing tests and implementing this method. But, one small part of the requirements for this particular task has not been met.

  Let's take another look at the requirement 👇🏼:
  
  > - Create a method called factorial that takes one argument, an integer N, and returns the [factorial](https://en.wikipedia.org/wiki/Factorial) for that number. The factorial is the multiplication of all integers from 1 to N and has the special case that the factorial of 0 is 1. This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.
  
  From the above requirement, the following edge case has not been covered in both the tests and the method implementation itself 👇🏼: 
  
  `This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.`
  
  Please make sure that the method only accepts `0` and positive numbers. Otherwise, ensure that the method raises an exception. This should also be covered with tests.
  
  Currently, the function returns `1` for negative numbers(`num < 0`), which shouldn't be true.

File: spec/factorial_spec.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      expect(solver.factorial(0)).to eq(1)
+    end
+
+    it 'returns 1 for 1' do
+      expect(solver.factorial(1)).to eq(1)
+    end
+
+    it 'returns 1 for negative numbers' do
+      expect(solver.factorial(-1)).to eq(1)
+    end
```
Comment (Lameck1) [inline spec/factorial_spec.rb]: - [ ] As discussed above, the requirements expect that the method raises an exception for negative numbers. It should not return `1` or any other value, for negative numbers. Please work on this.

**Review by Lameck1 — CHANGES_REQUESTED** [review]
Hi, @Dangelo-JAN, @TesheMaximillan,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Professional README ✔️
- Clean code ✔️ 
- Good commit messages ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
**Please, remember to tag me in your question so I can receive the notification.**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by descholar-ceo — APPROVED** [review]
Hi @Dangelo-JAN && @TesheMaximillan , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
